# grandTE
Genomic transposon prediction analysis





Third-party
-----------

grandTE package includes some third-party software:

* [blastn](https://ftp.ncbi.nlm.nih.gov/blast/executables/legacy.NOTSUPPORTED/)
* [muscle](http://www.drive5.com/muscle/downloads.htm)
* [mdust](https://github.com/lh3/mdust)
* [MITE-Hunter](http://target.iplantcollaborative.org/mite_hunter.html)
* [mustv2](http://www.healthinformaticslab.org/supp/resources.php)
* [mitetracker](https://github.com/INTABiotechMJ/MITE-Tracker)
* [MiteFinderII](https://github.com/jhu99/miteFinder)


References
-----------

Ge, Ruiquan et al. “MUSTv2: An Improved De Novo Detection Program for Recently Active Miniature Inverted Repeat Transposable Elements (MITEs).” Journal of integrative bioinformatics vol. 14,3 20170029. 10 Aug. 2017, [doi:10.1515/jib-2017-0029](https://pubmed.ncbi.nlm.nih.gov/28796642/)
Macko-Podgórni, Alicja et al. “A Global Landscape of Miniature Inverted-Repeat Transposable Elements in the Carrot Genome.” Genes vol. 12,6 859. 3 Jun. 2021, [doi:10.3390/genes12060859](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8227079/)
Crescente, Juan Manuel et al. “MITE Tracker: an accurate approach to identify miniature inverted-repeat transposable elements in large genomes.” BMC bioinformatics vol. 19,1 348. 3 Oct. 2018, [doi:10.1186/s12859-018-2376-y](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6171319/)
Ye, Congting et al. “detectMITE: A novel approach to detect miniature inverted repeat transposable elements in genomes.” Scientific reports vol. 6 19688. 22 Jan. 2016, [doi:10.1038/srep19688](https://pubmed.ncbi.nlm.nih.gov/26795595/)
Hu, Jialu et al. “MiteFinderII: a novel tool to identify miniature inverted-repeat transposable elements hidden in eukaryotic genomes.” BMC medical genomics vol. 11,Suppl 5 101. 20 Nov. 2018, [doi:10.1186/s12920-018-0418-y](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6245586/)


